[14.0][add] fieldservice_contract

hparfr opened the pull request:
Blocked by fieldservice_recurring <s>#762</s> https://github.com/OCA/field-service/pull/859
Blocked by fieldservice_quick_edit not published in pr yet: https://github.com/akretion/field-service/tree/14.0-mig-fieldservice_recurring_quick_edit

hparfr committed:
[add] fieldservice_contract

mourad-ehm commented on fieldservice_contract/models/contract_line.py:
```diff
@@ -0,0 +1,251 @@
+        """
+        for line in self:
+            # create order
+            if line.product_id.field_service_tracking == "order":
```
Field_service_tracking has not "order" value https://github.com/OCA/field-service/blob/14.0/fieldservice_sale/models/product_template.py#L17.  We have "sale" and "line" value which mean that we can create an unique FSM order by sale order if Field_service_tracking = "sale" or one SFM Order by sale line if Field_service_tracking = "line".
It seam that for recurring order we can only create one FSM recurring order by sale line.

mourad-ehm committed:
[FIX] rename field & fix bug field service tracking name

mourad-ehm committed:
[FIX] _prepare_invoice_line arg name

mourad-ehm committed:
[FIX] add fsm_location_id on sale line & nb of theoric fsm order by recurring order

mourad-ehm committed:
[FIX] compute price of realised fsm order

mourad-ehm committed:
[FIX] pre-commit format code

mourad-ehm committed:
[FIX] add fields in view

mourad-ehm committed:
[IMP] add invoice_policy: Planned fsm order smoothing bill,  realised order bill

mourad-ehm committed:
[FIX] compute _compute_tehoric_order_count

mourad-ehm committed:
[14.0][FIX] fieldservice_contract: field type

mourad-ehm committed:
[14.0][FIX] invoice policy help msg

mourad-ehm committed:
[14.0][FIX] fieldservice_contract: generate recuring based on line frequency

hparfr committed:
fieldservice_contract: adapt frequency to quickedit

paradoxxxzero committed:
[FIX] fieldservice_contract: Localize date to datetime conversion when creating a recurring order (#15)

hparfr committed:
fieldservice_contract: fix typo on date

mourad-ehm committed:
[FIX][14.0] fieldservic_contract: replace invoiceable_stage_ids field on contract by ._get_invoiceable_stage method

hparfr committed:
fieldservice_contract: hide by default advanced field

in the sale view

mourad-ehm committed:
[FIX][14.0] fieldservice_contract : date start & date end w'll be added by contract_invoice_start_end_dates

mourad-ehm committed:
[IMP][14.0] fieldservice_contract : simplify invoice_policy

mourad-ehm committed:
[IMP][14.0] fieldservice_contract : simplify invoice_policy

mourad-ehm committed:
[FIX][14.0] fieldservice_contract: del fsm_location_id from view (defined in other view)

mourad-ehm committed:
Del pdb

paradoxxxzero committed:
[IMP] fieldservice_contract Use new fieldservice_sale api

From https://github.com/OCA/field-service/pull/1049
Needs https://github.com/OCA/field-service/pull/1054

hparfr committed:
fieldservice_contract: fix default value to UTC

if self.env.user.tz is None then UTC

damdam-s committed:
[FIX] fix warning in odoo logs by changing the label

damdam-s committed:
remove 'inverse_name' param from Many2one declaration

petrus-v commented on fieldservice_contract/__manifest__.py:
```diff
@@ -0,0 +1,33 @@
+        "fieldservice_recurring",
+        "fieldservice_sale_recurring",
+        "fieldservice_sale_frequency",
+        "fieldservice_recurring_quick_edit",
```
I could not find any PR to this module fieldservice_recurring_quick_edit on version 14.0, does this dependency is required ? or could you point give me the link that I could give a review?

hparfr replied:
https://github.com/akretion/field-service/tree/14.0-mig-fieldservice_recurring_quick_edit

hparfr replied:
The dependency tree should be clarified / refactored.

First, we need `frequencies set` probably in his own module. Then defined in `sale order lines`, `contracts lines`.

`Frequency set` has two notions; 
- 1) abstract frequency set ( ambigus defined during the sale. Like "once a month".
- 2) concret frequency set, usually derived from the abstract. Like "first monday of each month at 6am".

fsm_recurring_quick_edit let you input an abstract frequency set` on the `sale order line` and convert it to a `concret frequency set` on a `recurring order`.

The abstract frequency set may be also used on the contract line.

petrus-v commented on fieldservice_contract/models/sale_order_line.py:
```diff
@@ -0,0 +1,60 @@
+        self, contract, predecessor_contract_line_id=False
+    ):
+        self.ensure_one()
+        contract.fsm_location_id = self.order_id.fsm_location_id
```
this is a little trap, as developer I won't expect the contract to be updated while preparing conctract liste, also it make hard to overload in submodules

mourad-ehm committed:
[FIX][14.0] fieldservice_contract: bug recuring creation (expected singleton)
